source/fe/mapping_q_cache.cc

```diff
@@ -125,6 +184,319 @@ MappingQCache<dim, spacedim>::initialize(
+
+              std::vector<Point<dim>> quadrature_points;
+              for (const auto i :
+                   FETools::hierarchic_to_lexicographic_numbering<dim>(
+                     this->polynomial_degree))
+                quadrature_points.push_back(quadrature_gl.point(i));
+              Quadrature<dim> quadrature(quadrature_points);
+
+              fe_values = std::make_unique<FEValues<dim, spacedim>>(
+                mapping, fe, quadrature, update_quadrature_points);
+            }
+
+          fe_values->reinit(cell);
+          points = fe_values->get_quadrature_points();
+        }
+
+      for (auto &p : points)
+        if (function_describes_relative_displacement)
+          p += transformation_function(cell, p);
+        else
+          p = transformation_function(cell, p);
+
+      return points;
+    });
+}
+
+
+
+template <int dim, int spacedim>
+void
+MappingQCache<dim, spacedim>::initialize(
+  const Mapping<dim, spacedim> &      mapping,
+  const Triangulation<dim, spacedim> &tria,
+  const Function<spacedim> &          transformation_function,
+  const bool                          function_describes_relative_displacement)
+{
+  AssertDimension(transformation_function.n_components, spacedim);
+
+  const std::function<Point<spacedim>(
+    const typename Triangulation<dim, spacedim>::cell_iterator &,
+    const Point<spacedim> &)>
+    fu = [&](const typename Triangulation<dim, spacedim>::cell_iterator &,
+             const Point<spacedim> &point) -> Point<spacedim> {
+    Point<spacedim> new_point;
+    for (int c = 0; c < spacedim; ++c)
+      new_point[c] = transformation_function.value(point, c);
+    return new_point;
+  };
+
+  this->initialize(mapping, tria, fu, function_describes_relative_displacement);
+}
+
+namespace
+{
+  template <typename VectorType>
+  void
+  import(const VectorType &vector,
+         LinearAlgebra::distributed::Vector<typename VectorType::value_type>
+           &vector_ghosted)
+  {
+    vector_ghosted.import(vector, VectorOperation::values::insert);
+  }
+
+  template <typename T>
+  void
+  import(const Vector<T> &                      vector,
+         LinearAlgebra::distributed::Vector<T> &vector_ghosted)
+  {
+    for (const auto i : vector_ghosted.locally_owned_elements())
+      vector_ghosted[i] = vector[i];
+  }
+
+  void
+  import(const TrilinosWrappers::MPI::Vector &vector,
+         LinearAlgebra::distributed::Vector<
+           typename TrilinosWrappers::MPI::Vector::value_type> &vector_ghosted)
+  {
+    for (const auto i : vector_ghosted.locally_owned_elements())
+      vector_ghosted[i] = vector[i];
+  }
+
+  void
+  import(const LinearAlgebra::EpetraWrappers::Vector &vector,
+         LinearAlgebra::distributed::Vector<
+           typename LinearAlgebra::EpetraWrappers::Vector::value_type>
+           &vector_ghosted)
+  {
+    Assert(false, ExcNotImplemented());
+    (void)vector;
+    (void)vector_ghosted;
+  }
+} // namespace
```

Comment:
An issue is how to get the input vectors (of arbitrary types) into the right shape. I used here the suggestion made in https://github.com/dealii/dealii/pull/11594#issuecomment-775774068: copy the values into `LA::d::Vector` with the right partitioner. I don't know if we want to discuss this issue here or we want to move it to #11594 since I would like to use the same approach there. I could instantiate the function for now only for `LA::d::Vector`. 

But in general, is there already an infrastructure to convert vectors?